var/spack/repos/builtin/packages/upcxx/package.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+import llnl.util.tty as tty
+import os
+
+
+class Upcxx(Package):
+    """UPC++ is a C++ library that supports Partitioned Global Address Space
+    (PGAS) programming, and is designed to interoperate smoothly and
+    efficiently with MPI, OpenMP, CUDA and AMTs. It leverages GASNet-EX to
+    deliver low-overhead, fine-grained communication, including Remote Memory
+    Access (RMA) and Remote Procedure Call (RPC)."""
+
+    homepage = "https://upcxx.lbl.gov"
+
+    version('2019.3.2', '844722cb0e8c0bc649017fce86469457')
+
+    variant('cuda', default=False, 
+            description='Builds a CUDA-enabled version of UPC++')
+    depends_on('cuda', when='+cuda')
+
+    def url_for_version(self, version):
+        if version > Version('2019.3.2'):
+            url = "https://bitbucket.org/berkeleylab/upcxx/downloads/upcxx-{0}.tar.gz"
+        else:
+            url = "https://bitbucket.org/berkeleylab/upcxx/downloads/upcxx-{0}-offline.tar.gz"
+        return url.format(version)
+
+    def setup_environment(self, spack_env, run_env):
+        if 'platform=cray' in self.spec:
+            if (os.environ.get('CROSS') is None):
```

Comment:
Would this be anticipated to work on Cray systems not at LBL?

Reply:
This is an excellent question, and I may need your help to handle this one. Basically, UPC++ provides a `upcxx-run` launcher to help new users getting all the right options set up (which are set via environment variables). Advanced users can set up the environment and then use the job spawner available on the platform they are using. However, because of this `upcxx-run` utility, it is important for UPC++ to know if the machine is using SLURM or ALPS. Valid values for the `CROSS` environment variable are `cray-aries-slurm` (NERSC Cori for instance) and `cray-aries-alps` (ALCF Theta) . When the CROSS environment *is not set*, the spack package will currently default it to `cray-aries-slurm` and display a warning message. Would this be an acceptable way to handle this case?

Reply:
Out of curiosity (not asking for a change) would it be possible to detect which of these apply rather than depending on an environment variable? Would you expect this environment variable is automatically set or would users normally need to set it for just this installation?

> However, because of this upcxx-run utility, it is important for UPC++ to know if the machine is using SLURM or ALPS

Does this mean the environment variable needs to be set at run time? FYI `spack_env` refers to the build time environment, so if you are using this function to generate a module, the module would not include this variable unless you set `run_env`.

> When the CROSS environment is not set, the spack package will currently default it to cray-aries-slurm and display a warning message.

If you use `tty.warn` it will display somewhat differently.

Reply:
> Out of curiosity (not asking for a change) would it be possible to detect which of these apply rather than depending on an environment variable?

Sadly this does not appear to be the case - all Cray XC systems ship with ALPS, but some sites install SLURM to use instead (and still others install SLURM for unrelated purposes). Therefore the existence of ALPS/SLURM binaries in PATH or variables in the environment is not a reliable indicator of which actually *work* for obtaining compute nodes at the site in question. The only way to know for sure is site documentation, or to actually try submitting a job and attempting to detect success (which suffers from the halting problem, and in any case seems outside the scope of what the build system should do).

The difficulty of detecting this in an automated way is why we ship two variants and ask the admin to select the correct one.